### PR TITLE
add OR_PATTERNS_BACK_COMPAT lint

### LINKED ISSUE #83318: Migration for or patterns
As part of the `|` patterns work ([tracking issue]), we plan to change the meaning of the `$x:pat` matcher in macros within the 2021 Edition. The current plan of record is:

* The `pat2015` matcher will match legacy patterns, which exclude the `|` character.
* The `pat2021` matcher will match new patterns, which include the `|` character.
* The `pat` matcher will be equivalent to `pat2021` from Rust 2021 forward, and equivalent to `pat2015` on older editions, since past crater runs (e.g., [1] and [2]) found significant and complex breakage from just changing `pat` to `pat2021` uniformly.

[tracking issue]: https://github.com/rust-lang/rust/issues/54883
[1]: https://github.com/rust-lang/rust/pull/78935#issuecomment-735050394
[2]: https://github.com/rust-lang/rust/pull/78935#issuecomment-740696077

The goal of this issue is to issue a useful lint that suggests that converting `$x:pat` to `$x:pat2015` if converting to Rust 2021 could cause breakage. We don't want to do this for *all* `$x:pat` usages, however, because the vast majority would benefit from being `$x:pat2021`. 

### Examples where lint should trigger

Here is an example of a macro that we DO want to convert ([source](https://github.com/JonasJebing/rust_match_any/blob/0fa72c02d4211b4edd513fc8d53f98a40caf9aca/src/lib.rs#L104-L113)):

```rust
#[macro_export]
macro_rules! match_any {
    ( $expr:expr , $( $( $pat:pat )|+ => $expr_arm:expr ),+ ) => {
        match $expr {
            $(
                $( $pat => $expr_arm, )+
            )+
        }
    };
}
```

Here is another example ([source](https://github.com/ratel-rust/ratel-core/blob/76f1b93e68cb6b0fadd89fddee70ea9084de3b55/core/src/tokenizer.rs#L13-L23)):

```rust
/// Helper macro for declaring byte-handler functions with correlating constants.
/// This becomes handy due to a lookup table present below.
macro_rules! define_handlers {
    { $(const $static_name:ident: $name:ident |$tok:pat, $byte:pat| $code:block)* } => {
        $(
            fn $name($tok: &mut Tokenizer, $byte: u8) -> Result<Token> $code

            const $static_name: fn(&mut Tokenizer, u8) -> Result<Token> = $name;
        )*
    }
}
```

### PR DESCRIPTION
**Pull request by 0xPoe**
close https://github.com/rust-lang/rust/issues/83318

### CODE REVIEW

**Comment by hi-rustin** [comment]
/hold

- [ ] Need to add comments
- [ ] Need to add tests

File: compiler/rustc_expand/src/mbe/macro_rules.rs [hunk 1]
```diff
@@ -953,6 +954,23 @@ fn check_matcher_core(
...
+                        if let TokenTree::Token(token) = next_token {
+                            if let BinOp(token) = token.kind {
+                                if let token::BinOpToken::Or = token {
+                                    sess.buffer_lint_with_diagnostic(
+                                        &OR_PATTERNS_BACK_COMPAT,
+                                        next_token.span(),
+                                        ast::CRATE_NODE_ID,
+                                        "detects usage of old versions of or patterns",
+                                        BuiltinLintDiagnostics::OrPatternsBackCompat(
+                                            "the pat2021 matcher will match new patterns, which include the | character. \
```
Comment (nikomatsakis) [inline compiler/rustc_expand/src/mbe/macro_rules.rs:967]:
```suggestion
                                            "in Rust 2021, the pat matcher will match new patterns, which include the | character. \
```

File: compiler/rustc_expand/src/mbe/macro_rules.rs [hunk 2]
```diff
@@ -953,6 +954,23 @@ fn check_matcher_core(
...
+                                Please use pat2015 to avoid breakage.".to_string()),
+                                    );
+                                }
+                            }
+                        }
+                    }
                     match is_in_follow(next_token, kind) {
```
Comment (nikomatsakis) [inline compiler/rustc_expand/src/mbe/macro_rules.rs:974]: We should modify the follow set for Pat2021 (do I understand correctly that, by the time we get to this code, we either have Pat2015 or Pat2021?) to exclude `|`, too.

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
This looks basically right! We do need tests though.

File: compiler/rustc_lint/src/context.rs [hunk 3]
```diff
@@ -707,6 +707,9 @@ pub trait LintContext: Sized {
                 BuiltinLintDiagnostics::ProcMacroBackCompat(note) => {
                     db.note(&note);
                 }
+                BuiltinLintDiagnostics::OrPatternsBackCompat(note) => {
+                    db.note(&note);
+                }
```
[the remark this answers is not in the document]
Comment (nikomatsakis) [reply compiler/rustc_lint/src/context.rs:712]: Something like:

```rust
db.span_suggestion(span, "use pat2015 to preserve semantics", "pat2015", Applicability::MachineApplicable);
```

actually you could probably just remove the `note`

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

File: compiler/rustc_lint/src/context.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (hi-rustin) [reply compiler/rustc_lint/src/context.rs:712]: Added, but I'm not sure if the span is in the right place?

[the remark this answers is not in the document]
Comment (hi-rustin) [reply compiler/rustc_lint/src/context.rs:712]: I actually don't know how to just get pat span, do you have any suggestions?

File: compiler/rustc_lint_defs/src/builtin.rs [hunk 4]
```diff
@@ -3133,3 +3134,37 @@ declare_lint! {
...
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// In Rust 2021, the pat matcher will match new patterns, which include the | character.
+    pub OR_PATTERNS_BACK_COMPAT,
+    Allow,
+    "detects usage of old versions of or patterns",
```
Comment (mark-i-m) [inline compiler/rustc_lint_defs/src/builtin.rs:3169]:
```suggestion
    "detects usage of old versions of or-patterns",
```

**Comment by mark-i-m** [comment]
I would also recommend adding a test for the following cases:
```
macro_rules! foo { ($x:pat | $x:pat) => {} } // emit suggestion
macro_rules! bar { ($(x:pat)+ | $($x:pat)+) => {} } // emit suggestion
macro_rules! baz { ($x:pat2015 | $x:pat2015) => {} } // should be ok
macro_rules! qux { ($x:pat2015 | $x:pat) => {} } // should be ok
macro_rules! ogg { ($x:pat | $x:pat2015) => {} } // emit suggestion
```

File: compiler/rustc_lint/src/context.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (hi-rustin) [reply compiler/rustc_lint/src/context.rs:714]: Added. Thank you for all your help!

**Comment by hi-rustin** [comment]
> I would also recommend adding a test for the following cases:
> 
> ```
> macro_rules! foo { ($x:pat | $x:pat) => {} } // emit suggestion
> macro_rules! bar { ($(x:pat)+ | $($x:pat)+) => {} } // emit suggestion
> macro_rules! baz { ($x:pat2015 | $x:pat2015) => {} } // should be ok
> macro_rules! qux { ($x:pat2015 | $x:pat) => {} } // should be ok
> macro_rules! ogg { ($x:pat | $x:pat2015) => {} } // emit suggestion
> ```

Added.

File: compiler/rustc_expand/src/mbe/macro_rules.rs [hunk 5]
```diff
@@ -951,8 +952,32 @@ fn check_matcher_core(
...
+                                    let suggestion = quoted_tt_to_string(&TokenTree::MetaVarDecl(
+                                        span,
+                                        name,
+                                        Some(NonterminalKind::Pat2015 { inferred: false }),
+                                    ));
+                                    sess.buffer_lint_with_diagnostic(
+                                        &OR_PATTERNS_BACK_COMPAT,
+                                        span,
+                                        ast::CRATE_NODE_ID,
+                                        &*format!("detects usage of old versions of or-patterns",),
```
Comment (nikomatsakis) [inline compiler/rustc_expand/src/mbe/macro_rules.rs:972]: We should change this message. Maybe "the meaning of the `pat` fragment specific is changing in Rust 2021, which may affect this macro"

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
One nit, but this looks great! Almost ready to land.

File: compiler/rustc_expand/src/mbe/macro_rules.rs (hunk 5 above)
Comment (hi-rustin) [reply compiler/rustc_expand/src/mbe/macro_rules.rs]: Changed.

**Review by nikomatsakis — APPROVED** [review]

File: compiler/rustc_expand/src/mbe/macro_rules.rs [hunk 6]
```diff
@@ -951,8 +952,32 @@ fn check_matcher_core(
...
+                                    let suggestion = quoted_tt_to_string(&TokenTree::MetaVarDecl(
+                                        span,
+                                        name,
+                                        Some(NonterminalKind::Pat2015 { inferred: false }),
+                                    ));
+                                    sess.buffer_lint_with_diagnostic(
+                                        &OR_PATTERNS_BACK_COMPAT,
+                                        span,
+                                        ast::CRATE_NODE_ID,
+                                        &*format!("the meaning of the pat fragment specific is changing in Rust 2021, which may affect this macro",),
```
Comment (mark-i-m) [inline compiler/rustc_expand/src/mbe/macro_rules.rs:972]:
```suggestion
                                        &*format!("the meaning of the `pat` fragment specificier is changing in Rust 2021, which may affect this macro",),
```

Minor typo...

Comment (nikomatsakis) [reply compiler/rustc_expand/src/mbe/macro_rules.rs:972]: actually, your correction is wrong too :)

Comment (nikomatsakis) [reply compiler/rustc_expand/src/mbe/macro_rules.rs:972]:
```suggestion
                                        &*format!("the meaning of the `pat` fragment specifier is changing in Rust 2021, which may affect this macro",),
```

**Comment by mark-i-m** [comment]
Oh, darn... the typo fix requires blessings.

**Comment by nikomatsakis** [comment]
@mark-i-m d'oh!

**Comment by nikomatsakis** [comment]
@hi-rustin can you bless the test please?

File: src/test/ui/macros/macro-or-patterns-back-compat.fixed [hunk 7]
```diff
@@ -0,0 +1,26 @@
+// ignore-tidy-linelength
+// run-rustfix
+
+#![feature(edition_macro_pats)]
+#![deny(or_patterns_back_compat)]
+#![allow(unused_macros)]
+macro_rules! foo { ($x:pat2015 | $y:pat) => {} } //~ ERROR the meaning of the pat fragment specific is changing in Rust 2021, which may affect this macro
```
Comment (nikomatsakis) [inline src/test/ui/macros/macro-or-patterns-back-compat.fixed:7]:
```suggestion
macro_rules! foo { ($x:pat2015 | $y:pat) => {} } //~ ERROR the meaning of the `pat` fragment specifier is changing in Rust 2021, which may affect this macro
```

File: src/test/ui/macros/macro-or-patterns-back-compat.stderr [hunk 8]
```diff
@@ -0,0 +1,32 @@
+error: the meaning of the pat fragment specific is changing in Rust 2021, which may affect this macro
```
Comment (nikomatsakis) [inline src/test/ui/macros/macro-or-patterns-back-compat.stderr:1]:
```suggestion
error: the meaning of the `pat` fragment specifier is changing in Rust 2021, which may affect this macro
```

File: src/test/ui/macros/macro-or-patterns-back-compat.stderr [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+LL | macro_rules! foo { ($x:pat | $y:pat) => {} }
+   |                     ^^^^^^ help: use pat2015 to preserve semantics: `$x:pat2015`
+   |
+note: the lint level is defined here
+  --> $DIR/macro-or-patterns-back-compat.rs:5:9
+   |
+LL | #![deny(or_patterns_back_compat)]
+   |         ^^^^^^^^^^^^^^^^^^^^^^^
+
+error: the meaning of the pat fragment specific is changing in Rust 2021, which may affect this macro
```
Comment (nikomatsakis) [inline src/test/ui/macros/macro-or-patterns-back-compat.stderr:13]:
```suggestion
error: the meaning of the `pat` fragment specifier is changing in Rust 2021, which may affect this macro
```

File: src/test/ui/macros/macro-or-patterns-back-compat.stderr [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+  --> $DIR/macro-or-patterns-back-compat.rs:8:23
+   |
+LL | macro_rules! bar { ($($x:pat)+ | $($y:pat)+) => {} }
+   |                       ^^^^^^ help: use pat2015 to preserve semantics: `$x:pat2015`
+
+error: the meaning of the pat fragment specific is changing in Rust 2021, which may affect this macro
```
Comment (nikomatsakis) [inline src/test/ui/macros/macro-or-patterns-back-compat.stderr:19]:
```suggestion
error: the meaning of the `pat` fragment specifier is changing in Rust 2021, which may affect this macro
```

File: src/test/ui/macros/macro-or-patterns-back-compat.stderr [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+  --> $DIR/macro-or-patterns-back-compat.rs:11:21
+   |
+LL | macro_rules! ogg { ($x:pat | $y:pat2015) => {} }
+   |                     ^^^^^^ help: use pat2015 to preserve semantics: `$x:pat2015`
+
+error: the meaning of the pat fragment specific is changing in Rust 2021, which may affect this macro
```
Comment (nikomatsakis) [inline src/test/ui/macros/macro-or-patterns-back-compat.stderr:25]:
```suggestion
error: the meaning of the `pat` fragment specifier is changing in Rust 2021, which may affect this macro
```

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

File: src/test/ui/macros/macro-or-patterns-back-compat.stderr [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+LL | macro_rules! ogg { ($x:pat | $y:pat2015) => {} }
+   |                     ^^^^^^ help: use pat2015 to preserve semantics: `$x:pat2015`
+
+error: the meaning of the `pat` fragment specifier is changing in Rust 2021, which may affect this macro
+  --> $DIR/macro-or-patterns-back-compat.rs:13:26
+   |
+LL |     ( $expr:expr , $( $( $pat:pat )|+ => $expr_arm:expr ),+ ) => {
+   |                          ^^^^^^^^ help: use pat2015 to preserve semantics: `$pat:pat2015`
+
+error: aborting due to 4 previous errors
```
Comment (mark-i-m) [inline src/test/ui/macros/macro-or-patterns-back-compat.stderr:31]:
```suggestion
error: aborting due to 4 previous errors

```

**Review by nikomatsakis — APPROVED** [review]

**Pull request merged**